fix: don't panic when logs waits for more than 5 seconds (v2)

martin-sucha opened the pull request:
<!-- Type of change
Please label this PR with one of the existing labels, depending on the scope of your change
-->

This should fix https://github.com/testcontainers/testcontainers-go/issues/946

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, etc.
-->

There are two commits.

The first commit fixes synchronization issues in producer tests.

The second commit removes panic when logs endpoint takes more than 5 seconds to respond. The panic happened at least with podman when no new logs appear when using follow and since parameters.

We keep retrying until the context is canceled (the retry request would fail anyway) or the producer is stopped, whichever comes first. This makes the retry behavior consistent. The HTTP calls to fetch logs are interrupted immediately (previously the producer waited for in-progress calls to complete).

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

Panicking the whole test suite breaks all my tests if the container ends early or there is some temporary error fetching the logs.

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes #946 
- Closes #606 
- Supersedes #947 

<!-- Recommended
## How to test this PR

Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->


<!-- Optional
## Follow-ups

Add here any thought that you consider could be identified as an actionable step once this PR is merged.
-->

mdelapenya commented on logconsumer_test.go:
```diff
@@ -31,7 +42,41 @@ func (g *TestLogConsumer) Accept(l Log) {
 	}
 
-	g.Msgs = append(g.Msgs, s)
+	g.mu.Lock()
```
Could we add a comment here explaining why the synchronisation must be done at the test level? You already added it in #947 :)

mdelapenya reviewed (commented):
I'm testing this PR with the same script as in #947:

```
for x in {1..20}; do go test -timeout 600s -run ^Test_StartStop$ github.com/testcontainers/testcontainers-go -count=1; done
```

And in general I still see certain flakiness, although different to the one described in that ticket: instead of receiving inconsistent messages, I get a `context deadline exceeded` error. 

Could you take a look at it?

Apart from that, the code at this moment LGTM

mdelapenya commented:
Hey @martin-sucha I came back from parental leave and I'm revisiting this PR. Did you saw my previous comments?

martin-sucha commented:
Hey @mdelapenya yes, thank you for the comments! I've been quite busy recently, this keeps falling through the cracks. I hope I can revisit this soon.

martin-sucha commented:
> I'm testing this PR with the same script as in #947:
> 
> ```
> for x in {1..20}; do go test -timeout 600s -run ^Test_StartStop$ github.com/testcontainers/testcontainers-go -count=1; done
> ```
> 
> And in general I still see certain flakiness, although different to the one described in that ticket: instead of receiving inconsistent messages, I get a `context deadline exceeded` error.
> 

I fixed the error handling so that context canceled is not treated as an error. It is merely a signal that the log producer should stop. Also updated the error handling for other errors to always retry until the log producer is stopped.

Tofel commented:
wdyt about a different approach:
* allow the caller to pass an optional timeout duration
* instead of panicking return an error if log producer cannot be started?

In this way caller would have bigger control over what it wants to do:
* retry
* pass the error higher up the call stack
* panic, etc

Why?:
* retrying until success might not always be what we want (what if it never succeeds and we retry it until test timeouts?)
* panicking inside a different goroutine makes it impossible to handle it in any way and some programs might be fine with running without logs

mdelapenya commented:
> wdyt about a different approach:
> 
> * allow the caller to pass an optional timeout duration
> * instead of panicking return an error if log producer cannot be started?
> 
> In this way caller would have bigger control over what it wants to do:
> 
> * retry
> * pass the error higher up the call stack
> * panic, etc
> 
> Why?:
> 
> * retrying until success might not always be what we want (what if it never succeeds and we retry it until test timeouts?)
> * panicking inside a different goroutine makes it impossible to handle it in any way and some programs might be fine with running without logs

@Tofel I like this. Would you be able to work on it? We are thinking about an eventual v1 version, so adding a breaking change in the signature/behaviour, if done properly, wouldn't be that bad.

Tofel commented:
@mdelapenya I will be able to work on that, yes, but I need to confirm with me team when. Will circle back once I know it.

mdelapenya commented:
> @mdelapenya it's me again... so my team is okay with me taking the time to propose an alternative solution (although I guess it will have to be a bit different, since we cannot return the error from that goroutine, but why not have an error channel on which the caller could listen in a dedicated goroutine?).

Indeed, I admit that we should totally rewrite the log consumer implementation for the v1, as I think it has grown in a non-organic manner.

> Before I do start with it, could you tell me when you're thinking that v1 might see the light of the day?

Not sure about it. I already have a draft doc with the things that I'd like to do/fix, which I'd make it public right after completed, in the form of a GH discussion. From there, I'll start creating issues in case the community wants to participate.

mdelapenya commented:
This PR has been superseded by #1971, please take a look at https://github.com/testcontainers/testcontainers-go/pull/1971#issuecomment-1885098616

I'd like to thank you for your time elaborating this PR. It has not been merged this time, but came with a lot of light for the final implementation. Thanks!
